clients/js/test/mpl_core/buy.test.ts

```diff
@@ -11,178 +11,535 @@ import {
+    collection: collectionA.address,
+    owner: traderA,
+    whitelist,
+    mintProof: mintProofA,
+  });
+  const depositNftIxB = await getDepositNftCoreInstructionAsync({
+    pool: poolB,
+    asset: nftB.address,
+    collection: collectionB.address,
+    owner: traderB,
+    whitelist,
+    mintProof: mintProofB,
   });
 
   await pipe(
-    await createDefaultTransaction(client, buyer),
+    await createDefaultTransaction(client, traderA),
+    (tx) => appendTransactionMessageInstruction(depositNftIxA, tx),
+    (tx) => appendTransactionMessageInstruction(depositNftIxB, tx),
+    (tx) => signAndSendTransaction(client, tx)
+  );
+
+  // Try to buy NFT B from pool A, the default derived NFT receipt will be from
+  // pool A and NFT B which won't exist so the error is account unintialized.
+  let buyNftIx = await getBuyNftCoreInstructionAsync({
+    rentPayer: payer.address,
+    owner: traderA.address,
+    pool: poolA,
+    taker: traderB,
+    asset: nftB.address,
+    collection: collectionB.address,
+    maxAmount: 1n,
+  });
+
+  let promise = pipe(
+    await createDefaultTransaction(client, traderB),
     (tx) => appendTransactionMessageInstruction(buyNftIx, tx),
     (tx) => signAndSendTransaction(client, tx)
   );
 
-  // NFT is now owned by the buyer.
-  t.like(await fetchAssetV1(client.rpc, asset.address), <
-    Account<AssetV1, Address>
-  >{
-    address: asset.address,
-    data: {
-      owner: buyer.address,
-    },
+  await expectCustomError(t, promise, ANCHOR_ERROR__ACCOUNT_NOT_INITIALIZED);
+
+  // Try to buy NFT B from pool A, but use the correct NFT receipt and token accounts so they exist.
+  // The error will then be a seeds constraint error because  the NFT receipt passed in doesn't match the
+  // one in the pool.
+
+  // Derive  NFT receipt
+  const [nftReceiptA] = await findNftDepositReceiptPda({
+    mint: nftA.address,
+    pool: poolA,
   });
 
-  // Pool is auto-closed by the last NFT being bought.
-  const maybePool = await fetchMaybePool(client.rpc, pool);
-  t.assert(!maybePool.exists);
+  buyNftIx = await getBuyNftCoreInstructionAsync({
+    rentPayer: payer.address,
+    owner: traderA.address,
+    pool: poolA,
+    taker: traderB,
+    asset: nftB.address,
+    collection: collectionB.address,
+    maxAmount: 1n,
+    nftReceipt: nftReceiptA,
+  });
 
-  // Fee vault balance increases.
-  const endingFeeVaultBalance = (await client.rpc.getBalance(feeVault).send())
-    .value;
-  t.assert(endingFeeVaultBalance > startingFeeVaultBalance);
+  promise = pipe(
+    await createDefaultTransaction(client, traderB),
+    (tx) => appendTransactionMessageInstruction(buyNftIx, tx),
+    (tx) => signAndSendTransaction(client, tx)
+  );
 
-  // Deposit Receipt is closed
-  await assertNftReceiptClosed({ t, client, pool, mint: asset.address });
+  await expectCustomError(t, promise, ANCHOR_ERROR__CONSTRAINT_SEEDS);
+
+  // Try to buy NFT A from pool B, the default derived NFT receipt will be from
```

Comment:
One regression test idea for all buys/sells:

We could abstract the IdlBuffer test flow + try to imitate deposit receipts to assert discriminator checks

Reply:
What's the advantage to doing this in addition to the fake pool tests with idlBuffer?